### PR TITLE
Split empty catch blocks with finally clauses.

### LINKED ISSUE #1029: Formatting of empty catch blocks can appear misleading
`dart format` really likes to collapse empty `{}` blocks onto a single line.  However, in the case of a `try`-`catch`-`finally` chain,  I think that it would be better not to do that if there is an non-empty block of statements later in the chain.  For example:

```dart
void main() {
  try {
    doSomething();
  } on Exception catch (e) {
  } finally {
    cleanupSomething();
  }
}
```
`dart format` (version 2.13.1 stable) formats it as:
```dart
void main() {
  try {
    doSomething();
  } on Exception catch (e) {} finally {
    cleanupSomething();
  }
}
```
which I think is rather misleading since it appears that `cleanupSomething()` is called only in response to a thrown `Exception`.

### PR DESCRIPTION
**Pull request by munificent**
Fix #1029

### CODE REVIEW

File: lib/src/nesting_level.dart [hunk 1]
```diff
@@ -25,7 +25,7 @@ class NestingLevel extends FastHash {
   /// The nesting level surrounding this one, or `null` if this is represents
   /// top level code in a block.
   NestingLevel? get parent => _parent;
-  NestingLevel? _parent;
+  final NestingLevel? _parent;
```
Comment (natebosch) [inline lib/src/nesting_level.dart:28]: Now that it is `final` you can make the field public and remove the getter.

File: lib/src/source_visitor.dart [hunk 2]
```diff
@@ -2034,7 +2022,6 @@ class SourceVisitor extends ThrowingAstVisitor {
     token(node.rightParenthesis);
     builder.unnest();
 
-    @override
```
Comment (natebosch) [inline lib/src/source_visitor.dart:2037]: How is this related to the rest of the CL?

Did the analyzer classes remove this method?

**Review by natebosch — APPROVED** [review]

File: lib/src/nesting_level.dart (hunk 1 above)
Comment (munificent) [reply lib/src/nesting_level.dart:28]: Ah, good catch. Done.

File: lib/src/source_visitor.dart (hunk 2 above)
Comment (munificent) [reply lib/src/source_visitor.dart:2037]: It's not related, I just noticed this in passing. This is a _local_ function. That metadata annotation never did anything useful. I must have accidentally copied it from a method at some point.

**Comment by jamesderlin** [comment]
Is this going to work only if there's a `finally` block?  Will something like:

```dart
void main() {
  try {
    doSomething();
  } on FooException {
  } on BarException {
    doSomething();
  }
}
```
be left alone, or will it be transformed to:

```dart
void main() {
  try {
    doSomething();
  } on FooException {} on BarException {
    doSomething();
  }
}
```
?

**Comment by munificent** [comment]
Ah, good question. I did not cover that case, but it would be good to do so. I'll take a stab at it.

**Comment by munificent** [comment]
Fix out: https://github.com/dart-lang/dart_style/pull/1042

**Pull request merged**